### PR TITLE
Add utils to calculate model complexity and adjust folder structure for mmcv/cnn/utils .

### PR DESCRIPTION
**Pull request by dreamerlin**
This PR adds  `flop_counter.py` to calculate model complexity like FLOPs and parameter counters. It also makes a directory `mmcv/cnn/utils` to keep useful util files for modules and puts `flop_counter.py` and `weight_init.py` into it.

### CODE REVIEW

File: mmcv/utils/flops_counter.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+                              _AvgPoolNd, _BatchNorm, _ConvNd,
+                              _ConvTransposeMixin, _MaxPoolNd)
+
+
+def get_model_complexity_info(model,
+                              input_res,
+                              print_per_layer_stat=True,
+                              as_strings=True,
+                              input_constructor=None,
+                              ost=sys.stdout):
```
Comment (hellock) [inline mmcv/utils/flops_counter.py:42]: Add a docstring.

File: mmcv/utils/flops_counter.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+            return str(round(flops / 10.**3, precision)) + ' ' + units
+        else:
+            return str(flops) + ' Mac'
+
+
+def params_to_string(params_num):
+    """converting number to string
+
+    :param float params_num: number
+    :returns str: number
```
Comment (hellock) [inline mmcv/utils/flops_counter.py:94]: Reformat the docstring.

File: mmcv/cnn/utils/flops_counter.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+    params_count = get_model_parameters_number(flops_model)
+    flops_model.stop_flops_count()
+
+    if as_strings:
+        return flops_to_string(flops_count), params_to_string(params_count)
+
+    return flops_count, params_count
+
+
+def flops_to_string(flops, units='GMac', precision=2):
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:95]: We can change `Mac` to `FLOPS`, resulting in `GFLOPS`, `MFLOPS`, etc.

File: mmcv/cnn/utils/flops_counter.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+    """Convert FLOPs number into a string.
+
+    Args:
+        flops (float): FLOPs number to be converted.
+        units (None | str): Converted FLOPs units. Options are None, "GMac",
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:100]: str | None

File: mmcv/cnn/utils/flops_counter.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+            return str(round(flops / 10.**9, precision)) + ' ' + units
+        elif units == 'MMac':
+            return str(round(flops / 10.**6, precision)) + ' ' + units
+        elif units == 'KMac':
+            return str(round(flops / 10.**3, precision)) + ' ' + units
+        else:
+            return str(flops) + ' Mac'
+
+
+def params_to_string(params_num):
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:128]: num_params

File: mmcv/cnn/utils/flops_counter.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+                                        _AvgPoolNd, _BatchNorm, _ConvNd,
+                                        _ConvTransposeMixin, _MaxPoolNd)
+
+
+def get_model_complexity_info(model,
+                              input_shape,
+                              print_per_layer_stat=True,
+                              as_strings=True,
+                              input_constructor=None,
+                              ost=sys.stdout):
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:42]: Use the same argument as in `print()`: `file=sys.stdout, flush= False`.

File: mmcv/cnn/utils/flops_counter.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+    """Get complexity information of a model.
+
+    This method can calculate FLOPs and parameter counts of a model with
+    corresponding input shape. It can also print complexity information for
+    each layer in a model.
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:47]: List supported layers in the docstring.

File: tests/test_flops_counter.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                                          conv_flops_counter_hook,
+                                          flops_to_string, params_to_string)
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+# yapf: disable
+gt_results = [{'model': 'nn.Linear(1024, 2)', 'input': (1, 1024), 'flops': 1.0, 'params': 2050.0},      # noqa: E501
```
Comment (hellock) [inline tests/test_flops_counter.py:16]: May add a break line after `[`:
```python
gt_results = [
    {},
    {},
]
```

File: tests/test_flops_counter.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.ELU())', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 224},  # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.PReLU())', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 225},    # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.AdaptiveMaxPool2d(8))', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 224},   # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.AdaptiveAvgPool2d(8))', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 224},   # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.AvgPool2d(2))', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 224},   # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.MaxPool2d(2))', 'input': (3, 16, 16), 'flops': 45472.0, 'params': 224},   # noqa: E501
+              {'model': 'nn.Sequential(nn.Conv2d(3, 8, 3), nn.Upsample((32, 32)))', 'input': (3, 16, 16), 'flops': 52096.0, 'params': 224}]    # noqa: E501
+# yapf: enable
```
Comment (hellock) [inline tests/test_flops_counter.py:39]: No need to use a string to represent models and use `eval()`.

**Comment by dreamerlin** [comment]
Update the scripts to the [latest one](https://github.com/sovrasov/flops-counter.pytorch/blob/master/ptflops/flops_counter.py), and remove some redundant codes which are not common for users.  
Also check the accuracy of the FLOPs and params output by comparing with the similar repo like [thop](https://github.com/Lyken17/pytorch-OpCounter) and [torchstat](https://github.com/Swall0w/torchstat).  
For some layers like `nn.UpSample` and `nn.Relu()` which are not supported in other repos, I checkout the codes related and make sure the calculation is right.

File: mmcv/cnn/utils/flops_counter.py [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+        as_strings (bool): Output FLOPs and params counts in a string form.
+            Default: True.
+        input_constructor (None | callable): If specified, it takes a callable
+            method that generates input. otherwise, it will generate a random
+            tensor with input shape to calculate FLOPs. Default: None.
+        flush (bool): same as that in :func:`print`. Default: False.
+        ost (stream): same as `file` param in :func:`print`.
+            Default: sys.stdout.
+
+    Returns:
```
Comment (hellock) [inline mmcv/cnn/utils/flops_counter.py:75]: tuple[float | str]

File: tests/test_flops_counter.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+from mmcv.cnn.utils.flops_counter import flops_to_string, params_to_string
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+# yapf: disable
+gt_results = [
+              {'model': nn.Conv1d(3, 8, 3), 'input': (3, 16), 'flops': 1120.0, 'params': 80.0},       # noqa: E501
```
Comment (hellock) [inline tests/test_flops_counter.py:15]: 4 whitespaces are enough.

File: tests/test_flops_counter.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        model, x, print_per_layer_stat=False)
+    assert flops == '0.17 GFLOPs' and params == str(656)
+
+    # test print per layer status
+    model = nn.Conv1d(3, 8, 3)
+    x = (3, 16)
+    out = StringIO()
+    get_model_complexity_info(model, x, ost=out)
+    assert out.getvalue(
+    ) == 'Conv1d(0.0 M, 100.000% Params, 0.0 GFLOPs, 100.000% FLOPs, 3, 8, kernel_size=(3,), stride=(1,))\n'  # noqa: E501
```
Comment (hellock) [inline tests/test_flops_counter.py:105]: To make it a little prettier,
```python
assert out.getvalue() == \
    'Conv1d(0.0 M, 100.000% Params, 0.0 GFLOPs, 100.000% FLOPs, 3, 8, kernel_size=(3,), stride=(1,))\n'  # noqa: E501
```
